src/core/src/pattern/op/optional.cpp

```diff
@@ -15,11 +15,17 @@ std::vector<ov::DiscreteTypeInfo> ov::pass::pattern::op::Optional::get_optional_
 bool ov::pass::pattern::op::Optional::match_value(Matcher* matcher,
                                                   const Output<Node>& pattern_value,
                                                   const Output<Node>& graph_value) {
-    ov::OutputVector or_in_values = input_values();
-    auto wrap_node = std::make_shared<ov::pass::pattern::op::WrapType>(optional_types, m_predicate, or_in_values);
-    or_in_values.push_back(wrap_node);
+    // Turn the Optional node into WrapType node to create a case where the Optional node is present
+    ov::OutputVector input_values_to_optional = input_values();
+    size_t num_input_values_to_optional = input_values_to_optional.size();
+    auto wrap_node =
+        std::make_shared<ov::pass::pattern::op::WrapType>(optional_types, m_predicate, input_values_to_optional);
 
-    if (matcher->match_value(std::make_shared<ov::pass::pattern::op::Or>(or_in_values), graph_value)) {
+    // Add the newly created WrapType node to the list containing its inputs and create an Or node with the list
+    input_values_to_optional.push_back(wrap_node);
```

Comment:
maybe we can make it clearer using something like that?
```
 // we should use only "data" (0th) input of pattern::Optional in Or
auto pattern = num_input_values_to_optional == 0? wrap_node :  std::make_shared<ov::pass::pattern::op::Or>({wrap_node, input_values_to_optional[0]});

if (matcher->match_value(pattern, ...
```

Reply:
and using std; and using ov::pass::pattern::op at the begging of file to simplify 

std::make_shared<ov::pass::pattern::op::Or> -> make_shared<Or>

Reply:
Fixed